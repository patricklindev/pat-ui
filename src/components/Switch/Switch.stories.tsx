// import { action } from '@storybook/addon-actions';
import React from 'react';
import Switch from './Switch';

export default {
  title: 'Switch',
  component: Switch,
};

export const DefaultSwitch = () => {
  const [showOn, setShowOn] = React.useState(false);
  const handleChange = () => {
    setShowOn(!showOn);
  };
  return (
    <Switch label="Default Switch" isChecked={showOn} onChange={handleChange} />
  );
};

export const DiffSizeSwitch = () => {
  const [showOn, setShowOn] = React.useState(false);
  return (
    <div>
      <Switch
        switchSize="sm"
        onChange={() => setShowOn(!showOn)}
        isChecked={showOn}
        label="Small Switch with default color"
      />

      <Switch switchSize="lg" label="Large Switch with default color" />
    </div>
  );
};

export const DiffColorSwitch = () => {
  const [showOnP, setShowOnP] = React.useState(false);
  const [showOnS, setShowOnS] = React.useState(false);

  console.log(showOnP);
  return (
    <div>
      <Switch
        switchColor="primary"
        onChange={() => {
          setShowOnP(!showOnP);
        }}
        label="Primary color Switch"
        isChecked={showOnP}
      />

      <Switch
        switchColor="secondary"
        onChange={() => setShowOnS(!showOnS)}
        isChecked={showOnS}
        label="Secondary color Switch"
      />
    </div>
  );
};

export const DiffTypeSwitch = () => {
  const [showOn, setShowOn] = React.useState(false);
  const [defaultOn, setDefaultOn] = React.useState(true);
  return (
    <div>
      <Switch
        onChange={() => {
          setShowOn(!showOn);
        }}
        isChecked={showOn}
        label="Default Switch"
      />

      <Switch label="switch with label" />

      <Switch
        onChange={() => {
          setDefaultOn(!defaultOn);
        }}
        isChecked={defaultOn}
        label="On as default switch"
      />

      <Switch
        disabled
        label="Defualt disabled switch"
        onChange={() => console.log('clicked')}
      />
    </div>
  );
};
