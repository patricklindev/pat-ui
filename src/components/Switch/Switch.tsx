import React, { FC, InputHTMLAttributes } from 'react';

import { classNames } from '../../utils/classNames';

export type SwitchSize = 'sm' | 'lg';
export type SwitchColor = 'primary' | 'secondary' | 'default';

export interface ISwitchProps {
  /**  set class name */
  className?: string;
  /**  set switch size */
  switchSize?: SwitchSize;
  /** set switch color */
  switchColor?: SwitchColor;
  /** set disabled switch */
  disabled?: boolean;
  /** set default to checked */
  isChecked?: boolean;
  /** set switch label */
  label?: string;
}

export type PatSwitchProps = ISwitchProps &
  InputHTMLAttributes<HTMLInputElement>;

/**
 * Switches toggle the state of a single setting on or off.
 *
 *  ```js
 * import {Switch} from 'pat-ui'
 * ```
 *
 */
const Switch: FC<PatSwitchProps> = (props) => {
  const {
    switchSize,
    className,
    disabled = false,
    switchColor,
    label,
    isChecked,
    ...rest
  } = props;
  let styleClasses = classNames('switch', {
    [`switch-${isChecked ? switchColor : 'default'}`]: true,
    [`switch-${switchSize}`]: !!switchSize,
    disabled: !!disabled,
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let swh = (
    <div className={styleClasses} data-testid="switch-element">
      <label className="switch-container">
        <div>
          <input
            disabled={disabled}
            type="checkbox"
            className="switch-toggle"
            data-testid="switch-toggle"
            checked={isChecked}
            {...rest}
          />
          <span
            //set class name based on if switch is disabled or not
            data-testid="test-slider"
            className={disabled ? 'disabled switch-slider' : 'switch-slider'}
          />
        </div>
      </label>
      <p className="switch-label">{label}</p>
    </div>
  );

  return swh;
};

Switch.defaultProps = {
  switchColor: 'default',
  disabled: false,
  switchSize: 'sm',
};

export default Switch;
