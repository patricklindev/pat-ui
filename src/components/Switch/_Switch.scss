.switch-label {
  display: flex;
  align-items: center;
  cursor: pointer;
  user-select: none;

  .switch-label-text {
    padding-left: 8px;
  }
}

.switch-input {
  display: none;
}

.switch {
  position: relative;
  display: inline-block;
  border-radius: 16px;
  transition: background-color 250ms ease;

  &.primary {
    background-color: #1976d2;

    &:hover {
      background-color: #0d47a1;
    }
  }

  &.secondary {
    background-color: #f50057;

    &:hover {
      background-color: #d17777;
    }
  }

  &.warning {
    background-color: #ff9800;

    &:hover {
      background-color: #f57c00;
    }
  }

  &.default {
    background-color: #bdbdbd;

    &:hover {
      background-color: #9e9e9e;
    }
  }

  &.small {
    width: 36px;
    height: 14px;
  }

  &.checked {
    background-color: #1976d2;
  }

  &.disabled {
    cursor: not-allowed;
  }
}

.thumb {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  border-radius: 50%;
  transition: left 250ms ease;
  box-shadow: 0 0 3px 3px rgba(0, 0, 0, 0.1);

  &.primary {
    background-color: #1976d2;
  }

  &.secondary {
    background-color: #d50000;
  }

  &.warning {
    background-color: #ff5722;
  }

  &.default {
    background-color: #616161;
  }

  &.small {
    left: 0;
    width: 18px;
    height: 18px;
  }

  &.checked {
    left: 18px;

    &.small {
      left: 14px;
    }
  }
}
