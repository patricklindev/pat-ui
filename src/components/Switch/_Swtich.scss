.switch {
  position: relative;
  display: inline-block;
  width: 60px;
  height: 20px;

  input {
    opacity: 0;
    width: 0;
    height: 0;
  }
  .label {
    display: inline-block;
    position: relative;
    left: 80px;
  }
  .slider {
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: #ccc;
    -webkit-transition: 0.4s;
    transition: 0.4s;

    &:before {
      position: absolute;
      content: '';
      height: 30px;
      width: 30px;
      left: 4px;
      bottom: -4px;
      background-color: rgb(127, 129, 131);
      -webkit-transition: 0.4s;
      transition: 0.4s;
    }
    &:hover:before {
      border: 4px solid rgba(222, 226, 230, 0.18);
      transition: 0;
    }
    &.round {
      border-radius: 34px;

      &:before {
        border-radius: 50%;
      }
    }
  }

  input:checked + .slider {
    background-color: rgba(222, 226, 230, 0.8);
  }
  input:checked + .slider:before {
    -webkit-transform: translateX(26px);
    -ms-transform: translateX(26px);
    transform: translateX(26px);
  }
}

.switch {
  &-small {
    top: 10px;
    width: 55px;
    height: 12px;
    .slider::before {
      height: 22px;
      width: 22px;
    }
  }
}

.switch {
  &-primary {
    .slider {
      &:before {
        background-color: rgba(30, 88, 247, 0.8);
      }
    }
    input:checked + .slider {
      background-color: rgba(30, 88, 247, 0.8);
    }
  }
}

.switch {
  &-secondary {
    .slider {
      &:before {
        background-color: rgba(243, 13, 75, 0.8);
      }
    }
    input:checked + .slider {
      background-color: rgba(243, 13, 75, 0.8);
    }
  }
}
