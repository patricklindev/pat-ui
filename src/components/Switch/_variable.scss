$switch-primary: rgba(30, 88, 247, 0.8);
$switch-secondary: rgba(243, 13, 75, 0.8);
