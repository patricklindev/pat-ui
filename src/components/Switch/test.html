<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<style>

.switch-comp {
	margin-top: 10px
}

.switch {
  position: relative;
  display: inline-block;
  width: 60px;
  height: 20px;
}

.switch input { 
  opacity: 0;
  width: 0;
  height: 0;
}

.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  -webkit-transition: .4s;
  transition: .4s;
}

.slider:before {
  position: absolute;
  content: "";
  height: 26px;
  width: 26px;
  left: -2px;
  bottom: -4px;
  background-color: #ccc;
  border: 1px solid black;
  -webkit-transition: .4s;
  transition: .4s;
}

input:checked + .slider {
  background-color: #2196F3;
}

input:focus + .slider {
  box-shadow: 0 0 1px #2196F3;
}

input:checked + .slider:before {
  -webkit-transform: translateX(34px);
  -ms-transform: translateX(34px);
  transform: translateX(34px);
}

/* Rounded sliders */
.slider.round {
  border-radius: 20px;
}

.slider.round:before {
  border-radius: 50%;
}

.want_content {
padding: 5px 2px;
height: 100%;
line-height: 100%;
margin: auto;
}

</style>
</head>
<body>
<div class="switch-comp">
<label class="switch">
  <input type="checkbox">
  <span class="slider round"></span>
</label>
<div class="switch-comp">

<div class="switch-comp">
<label class="switch">
  <input type="checkbox" checked id="want">
  <span class="slider round"></span> 
</label>
<label for="want" class="want-content">aaaa</label>
</div>

</body>
</html> 
