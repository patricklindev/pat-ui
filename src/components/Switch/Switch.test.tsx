import React from 'react';
import Switch, { PatSwitchProps } from './Switch';
import { render, fireEvent, screen } from '@testing-library/react';

describe('Switch', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Switch label="Snapshot Switch" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default switch', () => {
    render(<Switch label="Default Switch"></Switch>);
    const element = screen.getByTestId('switch-element');
    const checkbox = screen.getByRole('checkbox');
    const label = screen.getByText('Default Switch');
    const slider = screen.getByTestId('test-slider');

    expect(element).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(slider).toBeInTheDocument();
    expect(slider).toHaveClass('switch-slider');
    expect(element).toHaveClass('switch switch-sm');
    expect(checkbox).not.toBeChecked();
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
    expect(element).toHaveClass('switch-default');
  });

  it('should render corrsonding button based on props', () => {
    const switchLargePrimary: PatSwitchProps = {
      switchColor: 'primary',
      switchSize: 'lg',
      className: 'test',
    };

    render(
      <Switch label="Large Primary Switch" {...switchLargePrimary}></Switch>
    );
    const element = screen.getByTestId('switch-element');
    const checkbox = screen.getByRole('checkbox');
    const label = screen.getByText('Large Primary Switch');
    const slider = screen.getByTestId('test-slider');

    expect(element).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(slider).toBeInTheDocument();
    expect(slider).toHaveClass('switch-slider');
    expect(element).toHaveClass('switch switch-lg test');
    expect(checkbox).not.toBeChecked();
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
  });

  it('should render on switch', () => {
    const switchDefualtOn: PatSwitchProps = {
      isChecked: true,
    };

    render(<Switch label="Default On Switch" {...switchDefualtOn}></Switch>);
    const element = screen.getByTestId('switch-element');
    const checkbox = screen.getByRole('checkbox');
    const label = screen.getByText('Default On Switch');
    const slider = screen.getByTestId('test-slider');

    expect(element).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(slider).toBeInTheDocument();
    expect(slider).toHaveClass('switch-slider');
    expect(element).toHaveClass('switch switch-default switch-sm');
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
    expect(checkbox).toHaveAttribute('checked', '');
  });

  it('should render a disabled switch', () => {
    const disabledSwitchProp: PatSwitchProps = {
      disabled: true,
      onClick: jest.fn(),
    };
    render(<Switch label="Disabled Switch" {...disabledSwitchProp}></Switch>);
    const element = screen.getByTestId('switch-element');
    const checkbox = screen.getByRole('checkbox');
    const label = screen.getByText('Disabled Switch');
    const slider = screen.getByTestId('test-slider');
    expect(element).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(slider).toBeInTheDocument();
    expect(slider).toHaveClass('switch-slider');
    expect(element).toBeInTheDocument();
    expect(checkbox).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(slider).toBeInTheDocument();
    expect(element).toHaveClass('switch switch-default switch-sm');
    expect(disabledSwitchProp.onClick).toHaveBeenCalledTimes(0);
    fireEvent.click(element);
    expect(disabledSwitchProp.onClick).toHaveBeenCalledTimes(0);
    expect(slider).toHaveClass('disabled');
  });
});
