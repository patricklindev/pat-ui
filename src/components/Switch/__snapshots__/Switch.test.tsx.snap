// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="switch switch-default switch-sm"
    data-testid="switch-element"
  >
    <label
      class="switch-container"
    >
      <div>
        <input
          class="switch-toggle"
          data-testid="switch-toggle"
          type="checkbox"
        />
        <span
          class="switch-slider"
          data-testid="test-slider"
        />
      </div>
    </label>
    <p
      class="switch-label"
    >
      Snapshot Switch
    </p>
  </div>
</DocumentFragment>
`;
